tls-core/src/prf/slave.rs

```diff
@@ -1,131 +1,117 @@
     pub p2: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
 pub struct SlaveKe1 {
     /// H((ms xor opad) || H((ms xor ipad) || seed))
     pub a1: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
 pub struct SlaveKe2 {
     /// H((ms xor opad) || H((ms xor ipad) || a1))
     pub a2: [u8; 32],
 }
 
-pub struct SlaveCf {}
-
-impl PrfSlave<Initialized> {
-    pub fn new() -> Self {
-        Self { state: Initialized }
-    }
-
-    pub fn next(self, outer_hash_state: [u32; 8], m: MasterMs1) -> (SlaveMs1, PrfSlave<Ms1>) {
-        // H((pms xor opad) || H((pms xor ipad) || seed))
-        let a1 = finalize_sha256_digest(outer_hash_state.clone(), 64, &m.inner_hash);
-
-        (
-            SlaveMs1 { a1 },
-            PrfSlave {
-                state: Ms1 { outer_hash_state },
-            },
-        )
-    }
+#[derive(std::cmp::PartialEq, Copy, Clone)]
+pub enum PRFSlaveState {
+    Initialized,
+    SlaveMs1Sent,
+    SlaveMs2Sent,
+    SlaveMs3Sent,
+    SlaveKe1Sent,
+    SlaveKe2Sent,
 }
 
-impl PrfSlave<Ms1> {
-    pub fn next(self, m: MasterMs2) -> (SlaveMs2, PrfSlave<Ms2>) {
-        // H((pms xor opad) || H((pms xor ipad) || a1))
-        let a2 = finalize_sha256_digest(self.state.outer_hash_state.clone(), 64, &m.inner_hash);
-
-        (
-            SlaveMs2 { a2 },
-            PrfSlave {
-                state: Ms2 {
-                    outer_hash_state: self.state.outer_hash_state,
-                },
-            },
-        )
-    }
+#[derive(Copy, Clone)]
+pub struct PrfSlave {
+    state: PRFSlaveState,
+    // Depending on the state, the outer hash state will be used for master
+    // secret or for key expansion.
+    outer_hash_state: Option<[u32; 8]>,
 }
 
-impl PrfSlave<Ms2> {
-    pub fn next(self, m: MasterMs3) -> (SlaveMs3, PrfSlave<Ke1>) {
-        // H((pms xor opad) || H((pms xor ipad) || a2 || seed))
-        let p2 = finalize_sha256_digest(self.state.outer_hash_state, 64, &m.inner_hash);
-
-        (SlaveMs3 { p2 }, PrfSlave { state: Ke1 })
-    }
-}
-
-impl PrfSlave<Ke1> {
-    pub fn next(self, outer_hash_state: [u32; 8], m: MasterKe1) -> (SlaveKe1, PrfSlave<Ke2>) {
-        // H((pms xor opad) || H((pms xor ipad) || seed))
-        let a1 = finalize_sha256_digest(outer_hash_state.clone(), 64, &m.inner_hash);
-
-        (
-            SlaveKe1 { a1 },
-            PrfSlave {
-                state: Ke2 { outer_hash_state },
-            },
-        )
+impl PrfSlave {
+    pub fn new() -> Self {
+        Self {
+            state: PRFSlaveState::Initialized,
+            outer_hash_state: None,
+        }
     }
-}
-
-impl PrfSlave<Ke2> {
-    pub fn next(self, m: MasterKe2) -> (SlaveKe2, PrfSlave<Cf>) {
-        // H((pms xor opad) || H((pms xor ipad) || a1))
-        let a2 = finalize_sha256_digest(self.state.outer_hash_state.clone(), 64, &m.inner_hash);
 
-        (SlaveKe2 { a2 }, PrfSlave { state: Cf })
+    pub fn next(
+        &mut self,
+        message: PRFMessage,
+        outer_hash_state: Option<[u32; 8]>,
```

Comment:
Having this option in the function signature is clunky. Can we pass it in another way?